jaeger_client/senders.py

```diff
@@ -28,14 +31,71 @@
 
 
 class Sender(object):
-    def __init__(self, host, port, io_loop=None):
-        self.host = host
-        self.port = port
+    def __init__(self, io_loop=None):
+        from threading import Lock
         self.io_loop = io_loop or self._create_new_thread_loop()
+        self._process_lock = Lock()
+        self._process = None
+        self.spans = []
+
+    def append(self, span):
+        """Queue a span for subsequent submission calls to flush()"""
+        self.spans.append(span)
+
+    @tornado.gen.coroutine
+    def flush(self):
+        """
+        Examine span and process state before yielding to _flush() for batching and transport.
+        Returns a tuple for accounting the number of spans attempted to send along with any
+        raised Exception and error message info, if any, yielded from _flush().
+        """
+        if not self.spans:
+            return
+        with self._process_lock:
+            process = self._process
+            if not process:
+                return
+
+        num_spans, exc, error_msg = yield self._flush(self.spans, self._process)
+
+        self.spans = []
+
+        raise tornado.gen.Return((num_spans, exc, error_msg))
+
+    @tornado.gen.coroutine
+    def _flush(self, spans, process):
+        """
+        Batch spans and invokes send() while calculating spans sent and providing all exception
+        handling.  Override with specific batching logic and error messaging, if desired.
+        """
+        num_spans = len(spans)
+        exc = None
+        error_msg = ''
+
+        batch = thrift.make_jaeger_batch(spans=spans, process=process)
+
+        try:
+            yield self.send(batch)
+        except Exception as e:
+            exc = e
+            error_msg = 'Failed to send batch: %s'
 
+        raise tornado.gen.Return((num_spans, exc, error_msg))
```

Comment:
This is a special exception that allows coroutines to return values in older versions of python: http://www.tornadoweb.org/en/stable/gen.html#tornado.gen.Return.  It's necessary to get the exception info from the send generator.

Reply:
Ok, I got it. Still I'm not sure this is necessary. 

The only return value used is `num_spans`, which is everytime equal to `sender.span_count` with the actual implementation. 

This means that you can simply save a variable [here](https://github.com/ProvoK/jaeger-client-python/pull/1/files#diff-8a53213519652bcef509b8fbddbdfe86R184) (something like `num_spans = self._sender.span_count`) and then use it later for reporting the number of success/error spans.

Everything else (exceptions and theirs messages) can be handled normally, I suppose.

Reply:
Of course, makes more sense to fully adopt span_count and I've updated.  However, I did retain the `Sender._flush()` generated exception info so that transport-specific exception handling is possible in Sender and its subclasses, without having to maintain a complete, sender-aware try/except block in `Reporter._consume_queue()`.  Does that work for you?